File: apps/danse/engine/score.js
import { rand, pick } from "./rng.js";

const BED_DB = -5.0;
const VOICE_DB = -2.0;
const EVENT_DB = 0.0;
const SWELL_DB = -2.0;
const EVENT_KEEP = 0.125;
const EVENT_MAX_PER_SEC = 6;
const SEMITONES_AT_FAR = -12.0;
const SEMITONES_AT_NEAR = 4.0;
const CUTOFF_AT_FAR = 900.0;
const CUTOFF_AT_NEAR = 12000.0;
const VOICE_SECONDS = [5.0, 11.0];

function dbToGain(db) {
  return Math.pow(10, db / 20.0);
}

function depthTo(z, lo, hi) {
  const u = Math.max(0, Math.min(1, (z + 1.1) / 2.2));
  return lo + (hi - lo) * u;
}

export class ScoreEngine {
  constructor(bankUrl) {
    this.bankUrl = bankUrl;
    this.ctx = null;
    this.master = null;
    this.limiter = null;

    this.bank = null;
    this.audioBuffers = new Map();
    this.byKind = { bed: [], sustained: [], transient: [] };

    this.state = {
      voices: [],
      bedTimeout: null,
      lastEpoch: null,
      eventsThisSecond: 0,
      currentSecond: 0,
      previousFrameId: new Map(),
      previousCut: null,
      bedStopped: false
    };
  }

  initContext() {
    if (this.ctx) return;
    this.ctx = new (window.AudioContext || window.webkitAudioContext)();

    this.master = this.ctx.createGain();
    this.limiter = this.ctx.createDynamicsCompressor();
    this.limiter.threshold.value = -1.0;
    this.limiter.knee.value = 0.0;
    this.limiter.ratio.value = 20.0;
    this.limiter.attack.value = 0.005;
    this.limiter.release.value = 0.2;

    this.master.connect(this.limiter);
    this.limiter.connect(this.ctx.destination);

    // Safari unlock hack
    const osc = this.ctx.createOscillator();
    const gain = this.ctx.createGain();
    gain.gain.value = 0.001;
    osc.connect(gain);
    gain.connect(this.ctx.destination);
    osc.start(0);
    osc.stop(this.ctx.currentTime + 0.1);
  }

  async load() {
    const res = await fetch(this.bankUrl);
    this.bank = await res.json();
    for (const g of this.bank.grains) {
      if (!this.byKind[g.kind]) this.byKind[g.kind] = [];
      this.byKind[g.kind].push(g);
    }
    for (const pool of Object.values(this.byKind)) {
      pool.sort((a, b) => a.id.localeCompare(b.id));
    }
  }

  async resume() {
    this.initContext();
    if (this.ctx.state === "suspended") {
      await this.ctx.resume();
    }
  }

  async getAudio(grainId) {
    if (this.audioBuffers.has(grainId)) return this.audioBuffers.get(grainId);

    const url = new URL(`../public/bank/${grainId}.mp3`, window.location.href);
    try {
      const response = await fetch(url);
      const arrayBuffer = await response.arrayBuffer();
      const audioBuffer = await new Promise((resolve, reject) => {
        this.ctx.decodeAudioData(arrayBuffer, resolve, reject);
      });
      this.audioBuffers.set(grainId, audioBuffer);
      return audioBuffer;
    } catch (e) {
      console.warn(`Failed to load grain ${grainId}`, e);
      return null;
    }
  }

  choose(kind, seed, ...words) {
    const pool = this.byKind[kind] || [];
    if (!pool.length) return null;

    let toward = null;
    if (words.length > 0 && Array.isArray(words[words.length - 1])) {
      toward = words.pop();
    }

    if (!toward) return pick(pool, seed, ...words);

    const [axis, target] = toward;
    const values = pool.map(g => g[axis]);
    const mean = values.reduce((a, b) => a + b, 0) / values.length;
    const stdDev = Math.sqrt(values.reduce((a, b) => a + Math.pow(b - mean, 2), 0) / values.length);
    const scale = stdDev || 1.0;

    const weights = values.map(v => Math.exp(-Math.pow((v - target) / scale, 2)));
    const total = weights.reduce((a, b) => a + b, 0);

    if (total <= 0) return pick(pool, seed, ...words);

    const r = rand(seed, ...words) * total;
    let sum = 0;
    for (let i = 0; i < weights.length; i++) {
      sum += weights[i];
      if (sum >= r) return pool[i];
    }
    return pool[pool.length - 1];
  }

  playBuffer(buffer, options = {}) {
    if (!buffer || !this.ctx) {
      if (!this.ctx) console.warn("playBuffer: ctx is null");
      return null;
    }
    const source = this.ctx.createBufferSource();
    source.buffer = buffer;

    const gainNode = this.ctx.createGain();
    gainNode.gain.value = options.gain ?? 1.0;
    source.playbackRate.value = options.playbackRate || 1.0;

    source.connect(gainNode);
    gainNode.connect(this.ctx.destination);

    const now = this.ctx.currentTime;
    if (options.fade) {
      gainNode.gain.setValueAtTime(0, now);
      gainNode.gain.linearRampToValueAtTime(options.gain ?? 1.0, now + options.fade);
      if (options.duration) {
        gainNode.gain.setValueAtTime(options.gain ?? 1.0, now + options.duration - options.fade);
        gainNode.gain.linearRampToValueAtTime(0, now + options.duration);
      }
    } else if (options.duration) {
      source.stop(now + options.duration);
    }

    source.start(now);
    return { source, gain: gainNode };
  }

  async loopBed(seed, i = 0) {
    if (this.state.bedStopped) return;

    const g = pick(this.byKind.bed || [], seed, i, 601);
    if (!g) return;
    const buffer = await this.getAudio(g.id);
    if (!buffer) return;

    const pan = (rand(seed, i, 602) - 0.5) * 0.5;
    const duration = buffer.duration;

    this.playBuffer(buffer, {
      gain: dbToGain(BED_DB),
      pan,
      fadeIn: 1.2,
      fadeOut: 1.2
    });

    const overlap = 1.2;
    const nextWait = Math.max(duration - overlap, 1.0);
    this.state.bedTimeout = setTimeout(() => this.loopBed(seed, i + 1), nextWait * 1000);
  }

  startBed(seed) {
    this.state.bedStopped = false;
    this.loopBed(seed, 0);
  }

  stopBed() {
    this.state.bedStopped = true;
    if (this.state.bedTimeout) clearTimeout(this.state.bedTimeout);
  }

  async handleVoices(seed, t, cast, spread, dt, layoutInfo) {
    const nVoices = 8;

    const placed = cast.map(c => {
      const l = layoutInfo.get(c.id) || { z: 0, x: 0, opacity: 1, area: 0.1 };
      return { ...c, z: l.z, x: l.x, area: l.area, opacity: l.opacity };
    });
    placed.sort((a, b) => a.z - b.z);

    for (let slot = 0; slot < nVoices; slot++) {
      if (!this.state.voices[slot]) {
        this.state.voices[slot] = { take: 0, timer: 0 };
      }

      let v = this.state.voices[slot];
      v.timer -= dt;

      if (v.timer <= 0 && placed.length > 0) {
        const stride = placed.length / Math.min(nVoices, placed.length);
        const pIndex = Math.min(placed.length - 1, Math.floor(slot * stride));
        const p = placed[pIndex];

        const span = VOICE_SECONDS[0] + (VOICE_SECONDS[1] - VOICE_SECONDS[0]) * rand(seed, slot, v.take, 701);
        v.timer = span;
        v.take++;

        const semis = depthTo(p.z, SEMITONES_AT_FAR, SEMITONES_AT_NEAR);
        const cutoff = depthTo(p.z, CUTOFF_AT_FAR, CUTOFF_AT_NEAR);
        const targetCentroid = depthTo(p.z, 350.0, 1800.0);

        const g = this.choose("sustained", seed, slot, v.take, 702, ["centroid", targetCentroid]);
        if (g) {
          this.getAudio(g.id).then(buffer => {
            if (!buffer) return;
            const playbackRate = Math.pow(2, semis / 12.0);
            const width = 0.15 + 0.85 * Math.min(1.0, spread * 1.6);
            const gain = dbToGain(VOICE_DB) * (0.45 + 0.55 * p.opacity) * (0.6 + 0.4 * Math.min(1.0, p.area * 40));

            this.playBuffer(buffer, {
              gain,
              pan: p.x * width,
              playbackRate,
              cutoff,
              fadeIn: 0.35,
              fadeOut: 0.9,
              duration: span
            });
          });
        }
      }
    }
  }

  async handleEvents(seed, t, cast, spread, cut, layoutInfo) {
    const sec = Math.floor(t);
    if (this.state.currentSecond !== sec) {
      this.state.currentSecond = sec;
      this.state.eventsThisSecond = 0;
    }

    // cast structure: { id, rect, layers: [{frame, weight}] }
    // A frame change in the primary layer (layer[0]) is a recast
    const nowMap = new Map(cast.map(c => [c.id, c.layers?.[0]?.frame]));

    if (this.state.previousFrameId.size > 0 && cut === this.state.previousCut) {
      let index = Math.floor(t * 100);

      for (const [id, frame] of nowMap.entries()) {
        if (this.state.previousFrameId.has(id) && this.state.previousFrameId.get(id) !== frame) {
          index++;
          const l = layoutInfo.get(id);
          if (!l) continue;

          if (rand(seed, index, 801) >= EVENT_KEEP) continue;
          if (this.state.eventsThisSecond >= EVENT_MAX_PER_SEC) continue;
          this.state.eventsThisSecond++;

          const z = l.z;
          const targetCentroid = depthTo(z, 500.0, 2200.0);
          const g = this.choose("transient", seed, index, 802, ["centroid", targetCentroid]);
          if (g) {
            this.getAudio(g.id).then(buffer => {
              if (!buffer) return;
              const playbackRate = Math.pow(2, depthTo(z, -7.0, 3.0) / 12.0);
              const cutoff = depthTo(z, 1400.0, 14000.0);
              const width = 0.2 + 0.8 * Math.min(1.0, spread * 1.6);
              const gain = dbToGain(EVENT_DB) * (0.5 + 0.5 * Math.min(1.0, l.area * 60));

              this.playBuffer(buffer, {
                gain,
                pan: l.x * width,
                playbackRate,
                cutoff,
                fadeIn: 0.001,
                fadeOut: 0.03
              });
            });
          }
        }
      }
    }
    this.state.previousFrameId = nowMap;
    this.state.previousCut = cut;
  }

  async handleReseed(seed, t, epoch) {
    if (this.state.lastEpoch === null) {
      this.state.lastEpoch = epoch;
      this.state.reseedK = 0;
      return;
    }

    if (epoch !== this.state.lastEpoch) {
      this.state.lastEpoch = epoch;
      const k = this.state.reseedK++;

      const g = pick(this.byKind.bed || [], seed, k, 901);
      if (g) {
        this.getAudio(g.id).then(buffer => {
          if (!buffer) return;
          const playbackRate = Math.pow(2, -24.0 / 12.0);
          const seconds = Math.max(1.6, 6.0 * Math.pow(0.72, k));

          this.playBuffer(buffer, {
            gain: dbToGain(SWELL_DB),
            pan: 0.0,
            playbackRate,
            fadeIn: seconds * 0.1,
            fadeOut: 0.25,
            duration: seconds
          });
        });
      }
    }
  }

  update(r, seed, t, layoutInfo) {
    if (!this.ctx) return;
    // Do NOT gate on ctx.state === "running": after ctx.resume() the state
    // may stay "suspended" for several frames on Chrome/Firefox. Queued
    // source.start() calls execute once the context actually resumes, so
    // scheduling while still suspended is correct and necessary.

    const dt = this.state.lastT !== undefined ? Math.max(0, t - this.state.lastT) : 0;
    this.state.lastT = t;

    const { state, cast } = r;
    this.handleVoices(seed, t, cast, state.spread, dt, layoutInfo);
    this.handleEvents(seed, t, cast, state.spread, state.cut, layoutInfo);
    this.handleReseed(seed, t, state.epoch);
  }
}

File: apps/danse/sound/score.py
#!/usr/bin/env python3
"""f(seed, program) -> stereo WAV. The same seed as the picture.

The score is not written to the film; it is DERIVED from it. `control.mjs` runs
the real engine over the real program and reports what the picture is doing —
how deep each plane hangs, where it sits across the frame, which cells swapped
photograph — and this turns that into sound using only grains cut from the two
recordings confirmed to be from that apartment.

    bed        the room's air, continuous, crossfaded end to end. It is under
               everything because the room was under everything.
    voices     eight planes sampled through the depth order. Depth sets pitch
               and darkness, lateral position sets pan, so the chord IS the
               spatial arrangement — when the camera moves and the depth order
               changes, the chord voices itself differently.
    events     a cell swapping photograph fires a transient, DECIMATED to about
               one in eight. Throwing most of them away is the single decision
               that makes the rhythm countable by a dancer rather than by a
               profiler: at 30 re-casts a second, keeping them all is a wash.
    reseed     each epoch turn takes a bed grain down two octaves and swells it,
               then cuts to silence. Shortening across the accelerating reseeds.

Two things this deliberately does NOT do:

  * No synthesis. There is no oscillator anywhere. The sub-bass under a reseed is
    the room itself at quarter speed. If it is in the film, it was in the air of
    that apartment in 2017.
  * No artificial reverb. The material was recorded IN the room, so it already
    carries the room; convolving it with a synthetic hall would only smear what
    is already true. Distance is rendered the way distance actually reaches an
    ear — quieter, darker (air absorbs treble), lower, and narrower — which is
    filtering of the room's own sound rather than an invention of a second space.

    apps/danse/sound/score.py                       # the 6:30 master
    apps/danse/sound/score.py --window trailer
    apps/danse/sound/score.py --seed 0x5F1E --window reel
"""

from __future__ import annotations

import argparse
import json
import math
import subprocess
import sys
from pathlib import Path

import numpy as np
from scipy.io import wavfile
from scipy.signal import lfilter, resample_poly

HERE = Path(__file__).resolve().parent
sys.path.insert(0, str(HERE))
BANK = HERE / "bank"
CONTROL = HERE / "control.mjs"
OUT = HERE / "out"

SR = 48_000

# Delivery. Soundscape Park is outdoors and Midnight Moment is Times Square, so
# the true-peak ceiling is not decorative — a limiter downstream that finds an
# inter-sample overshoot will pump the whole bed.
TARGET_LUFS = -16.0
TRUE_PEAK_DBTP = -1.0

# One event in eight. See the module docstring: this is a compositional decision
# with a number attached, not a performance budget.
EVENT_KEEP = 0.125
# And no more than this many in any one second, whatever the picture is doing.
EVENT_MAX_PER_SEC = 6

# How far a plane's depth moves it. `z` runs about -1.1 to +1.1; NEGATIVE IS
# FARTHER (the camera looks down -Z), so a far plane is slowed, darkened and
# pulled toward the middle.
SEMITONES_AT_FAR = -12.0
SEMITONES_AT_NEAR = 4.0
CUTOFF_AT_FAR = 900.0
CUTOFF_AT_NEAR = 12_000.0

# A voice holds its grain this long before re-picking, jittered per slot. Shorter
# and it chatters; longer and it stops tracking the arrangement.
VOICE_SECONDS = (5.0, 11.0)

BED_DB = -20.0
VOICE_DB = -21.0
EVENT_DB = -17.0
SWELL_DB = -14.0


# ── the same hash the picture uses ─────────────────────────────────────────────
# `rng.py` is a value-for-value port of engine/rng.js, so a sound decision and a
# picture decision made from the same seed and the same words agree. It lives in
# its own dependency-free module because `check-danse.py` holds the two
# implementations to identical values on every run.

from rng import pick, rand  # noqa: E402


# ── the bank ───────────────────────────────────────────────────────────────────


class Bank:
    """The grains, and the index that lets one be chosen without hearing it."""

    def __init__(self, root: Path):
        index = root / "bank.json"
        if not index.exists():
            sys.exit(f"no grain bank at {index} — run apps/danse/sound/1_bank.py first")
        data = json.loads(index.read_text())
        if data.get("rate") != SR:
            sys.exit(f"bank is {data.get('rate')} Hz, the score renders at {SR}")
        self.root = root
        self.fingerprint = data.get("fingerprint", "")
        self.sources = data.get("sources", [])
        self.grains = data["grains"]
        self.by_kind: dict[str, list[dict]] = {}
        for g in self.grains:
            self.by_kind.setdefault(g["kind"], []).append(g)
        for pool in self.by_kind.values():
            pool.sort(key=lambda g: g["id"])  # order must not depend on the filesystem
        self._audio: dict[str, np.ndarray] = {}

    def audio(self, grain: dict) -> np.ndarray:
        got = self._audio.get(grain["id"])
        if got is None:
            _, got = wavfile.read(self.root / f"{grain['id']}.wav")
            got = got.astype(np.float64)
            self._audio[grain["id"]] = got
        return got

    def choose(self, kind: str, *words: int, toward: tuple[str, float] | None = None) -> dict | None:
        """One grain, deterministically.

        `toward` biases the draw along an index axis — ("centroid", 300) prefers
        dark material. This is the same shape as `corpus.choose(candidates, ...)`
        on the picture side: weight the pool, then draw from it with the seed.
        """
        pool = self.by_kind.get(kind, [])
        if not pool:
            return None
        if toward is None:
            return pick(pool, *words)
        axis, target = toward
        values = np.array([g[axis] for g in pool], dtype=float)
        scale = float(np.std(values)) or 1.0
        weights = np.exp(-(((values - target) / scale) ** 2))
        total = float(weights.sum())
        if total <= 0:
            return pick(pool, *words)
        r = rand(*words) * total
        idx = int(np.searchsorted(np.cumsum(weights), r))
        return pool[min(idx, len(pool) - 1)]


# ── shaping ────────────────────────────────────────────────────────────────────


def varispeed(x: np.ndarray, semitones: float) -> np.ndarray:
    """Pitch by resampling — tape, not a phase vocoder.

    Duration moves with pitch, which is wanted: a plane hanging further away is
    both lower and longer, the way a slowed recording is. A phase vocoder would
    hold the duration and add the smearing that always comes with it, and there
    is no musical reason here to pay that.
    """
    ratio = 2 ** (semitones / 12.0)
    if abs(ratio - 1.0) < 1e-3 or len(x) < 8:
        return x
    up, down = 1000, max(1, int(round(1000 * ratio)))
    return resample_poly(x, up, down)


def darken(x: np.ndarray, cutoff: float) -> np.ndarray:
    """One-pole lowpass — air absorbing treble over distance."""
    if cutoff >= SR / 2.5 or len(x) < 4:
        return x
    a = math.exp(-2.0 * math.pi * cutoff / SR)
    return lfilter([1.0 - a], [1.0, -a], x)


def fades(n: int, head: float, tail: float) -> np.ndarray:
    win = np.ones(n)
    a, b = min(int(head * SR), n // 2), min(int(tail * SR), n // 2)
    if a:
        win[:a] = np.sin(np.linspace(0, np.pi / 2, a)) ** 2
    if b:
        win[n - b :] = np.sin(np.linspace(np.pi / 2, 0, b)) ** 2
    return win


def place(buf: np.ndarray, mono: np.ndarray, at: int, gain: float, pan: float) -> None:
    """Sum one mono signal into the stereo buffer at a sample offset.

    Equal-power pan, and a width that closes as the plane recedes: `pan` is
    already scaled by the caller, so a far plane arrives near the centre.
    """
    if at >= buf.shape[1] or len(mono) == 0:
        return
    n = min(len(mono), buf.shape[1] - at)
    theta = (np.clip(pan, -1.0, 1.0) + 1.0) * (math.pi / 4)
    buf[0, at : at + n] += mono[:n] * gain * math.cos(theta)
    buf[1, at : at + n] += mono[:n] * gain * math.sin(theta)


def db_to_gain(x: float) -> float:
    return 10 ** (x / 20.0)


# ── loudness, per BS.1770 ──────────────────────────────────────────────────────


def k_weight(x: np.ndarray) -> np.ndarray:
    """The two-stage K filter: a shelf, then a high-pass. Coefficients at 48 kHz."""
    shelf_b = [1.53512485958697, -2.69169618940638, 1.19839281085285]
    shelf_a = [1.0, -1.69065929318241, 0.73248077421585]
    hp_b = [1.0, -2.0, 1.0]
    hp_a = [1.0, -1.99004745483398, 0.99007225036621]
    return lfilter(hp_b, hp_a, lfilter(shelf_b, shelf_a, x))


def lufs(stereo: np.ndarray) -> float:
    """Integrated loudness with the -10 LU relative gate."""
    weighted = np.array([k_weight(ch) for ch in stereo])
    block, step = int(0.400 * SR), int(0.100 * SR)
    if weighted.shape[1] < block:
        return -math.inf
    starts = range(0, weighted.shape[1] - block + 1, step)
    power = np.array([float((weighted[:, s : s + block] ** 2).mean(axis=1).sum()) for s in starts])
    loud = -0.691 + 10 * np.log10(np.maximum(power, 1e-12))

    keep = loud > -70.0  # absolute gate
    if not keep.any():
        return -math.inf
    relative = -0.691 + 10 * np.log10(power[keep].mean()) - 10.0
    keep &= loud > relative  # relative gate
    if not keep.any():
        return -math.inf
    return float(-0.691 + 10 * np.log10(power[keep].mean()))


def true_peak_db(stereo: np.ndarray) -> float:
    """4x oversampled peak — what a downstream converter will actually see."""
    up = np.array([resample_poly(ch, 4, 1) for ch in stereo])
    return 20 * math.log10(max(float(np.abs(up).max()), 1e-12))


# ── the score ──────────────────────────────────────────────────────────────────


def depth_to(z: float, lo: float, hi: float) -> float:
    """Map a plane's signed depth into a range. Negative z is farther."""
    u = float(np.clip((z + 1.1) / 2.2, 0.0, 1.0))
    return lo + (hi - lo) * u


def render(control: dict, bank: Bank, quiet: bool = False) -> np.ndarray:
    seed = int(control["seed"])
    rate = float(control["rate"])
    frames = control["frames"]
    total = int(round(control["duration"] * SR))
    buf = np.zeros((2, total), dtype=np.float64)
    counts = {"bed": 0, "voice": 0, "event": 0, "swell": 0}

    # ── bed ────────────────────────────────────────────────────────────────────
    # Crossfaded end to end for the whole window. The one continuous thing.
    bed_pool = bank.by_kind.get("bed", [])
    if bed_pool:
        overlap = int(1.2 * SR)
        at = 0
        i = 0
        while at < total:
            g = pick(bed_pool, seed, i, 601)
            a = bank.audio(g)
            a = a * fades(len(a), 1.2, 1.2)
            # Alternate the pan so successive grains open the image rather than
            # stacking dead centre.
            place(buf, a, at, db_to_gain(BED_DB), (rand(seed, i, 602) - 0.5) * 0.5)
            at += max(len(a) - overlap, SR)
            i += 1
            counts["bed"] += 1

    # ── voices ─────────────────────────────────────────────────────────────────
    # Each slot holds a grain for a while, then re-picks. It takes the depth at
    # the moment it picks, so a voice is a photograph of the arrangement, held.
    n_voices = int(control.get("voices", 8))
    for slot in range(n_voices):
        t = 0.0
        take = 0
        while t < control["duration"]:
            fi = min(len(frames) - 1, int(t * rate))
            frame = frames[fi]
            span = VOICE_SECONDS[0] + (VOICE_SECONDS[1] - VOICE_SECONDS[0]) * rand(seed, slot, take, 701)
            voices = frame["v"]
            if not voices or frame["cut"] == "black":
                t += span
                take += 1
                continue
            z, opacity, area, x = voices[min(slot, len(voices) - 1)]

            semis = depth_to(z, SEMITONES_AT_FAR, SEMITONES_AT_NEAR)
            cutoff = depth_to(z, CUTOFF_AT_FAR, CUTOFF_AT_NEAR)
            # Far planes are dark, so ask the index for dark material and then
            # darken it further — selection first, filtering second.
            g = bank.choose("sustained", seed, slot, take, 702, toward=("centroid", depth_to(z, 350.0, 1800.0)))
            if g is None:
                break
            a = darken(varispeed(bank.audio(g), semis), cutoff)
            a = a * fades(len(a), 0.35, 0.9)

            # A flat arrangement (spread 0) is the 2017 composite: one picture
            # plane, no depth to voice. The chord closes to a single centred note.
            width = 0.15 + 0.85 * min(1.0, frame["sp"] * 1.6)
            gain = db_to_gain(VOICE_DB) * (0.45 + 0.55 * opacity) * (0.6 + 0.4 * min(1.0, area * 40))
            place(buf, a, int(t * SR), gain, x * width)
            counts["voice"] += 1
            t += span
            take += 1

    # ── events ─────────────────────────────────────────────────────────────────
    kept_in_second: dict[int, int] = {}
    index = 0
    for frame in frames:
        for z, area, x in frame["e"]:
            index += 1
            if rand(seed, index, 801) >= EVENT_KEEP:
                continue
            second = int(frame["t"])
            if kept_in_second.get(second, 0) >= EVENT_MAX_PER_SEC:
                continue
            kept_in_second[second] = kept_in_second.get(second, 0) + 1

            g = bank.choose("transient", seed, index, 802, toward=("centroid", depth_to(z, 500.0, 2200.0)))
            if g is None:
                break
            a = darken(varispeed(bank.audio(g), depth_to(z, -7.0, 3.0)), depth_to(z, 1400.0, 14_000.0))
            a = a * fades(len(a), 0.001, 0.03)
            width = 0.2 + 0.8 * min(1.0, frame["sp"] * 1.6)
            gain = db_to_gain(EVENT_DB) * (0.5 + 0.5 * min(1.0, area * 60))
            place(buf, a, int(frame["t"] * SR), gain, x * width)
            counts["event"] += 1

    # ── reseed ─────────────────────────────────────────────────────────────────
    # The room at quarter speed, swelling, then cut. Not a synthesised sub — the
    # same air, two octaves down.
    turns = []
    last_epoch = frames[0]["ep"] if frames else 0
    for frame in frames:
        if frame["ep"] != last_epoch:
            turns.append(frame["t"])
            last_epoch = frame["ep"]
    for k, when in enumerate(turns):
        g = bank.choose("bed", seed, k, 901)
        if g is None:
            break
        a = varispeed(bank.audio(g), -24.0)
        # Shortening across the accelerating reseeds — the film compresses, and
        # the sound has to compress with it or it reads as a mistake.
        seconds = max(1.6, 6.0 * (0.72**k))
        a = a[: int(seconds * SR)]
        a = a * np.linspace(0.0, 1.0, len(a)) ** 2 * fades(len(a), 0.05, 0.25)
        place(buf, a, max(0, int((when - seconds * 0.82) * SR)), db_to_gain(SWELL_DB), 0.0)
        counts["swell"] += 1

    if not quiet:
        print(
            f"  {counts['bed']} bed · {counts['voice']} voice notes · "
            f"{counts['event']}/{index} events kept · {counts['swell']} reseed swells"
        )
    return buf


def limit(stereo: np.ndarray, ceiling_db: float, lookahead: float = 0.005, release: float = 0.20) -> np.ndarray:
    """Look-ahead peak limiter, channels linked so the image does not wander.

    A plain attenuation cannot satisfy both delivery numbers at once: pulling the
    whole file down to clear -1 dBTP took the first render from -16 to -18 LUFS,
    and turning it back up puts the peaks straight back. The peaks have to come
    down WITHOUT the body coming with them, which is what a limiter is.

    Gain reduction is computed per sample, taken as a running minimum across the
    look-ahead window (so the reduction is already in place before the transient
    arrives — that is the attack), then released with a one-pole so it recovers
    smoothly rather than pumping on every hit.
    """
    from scipy.ndimage import minimum_filter1d

    ceiling = db_to_gain(ceiling_db)
    peak = np.abs(stereo).max(axis=0)
    if peak.max() <= ceiling:
        return stereo

    need = np.minimum(1.0, ceiling / np.maximum(peak, 1e-12))
    w = max(1, int(lookahead * SR))
    need = minimum_filter1d(need, size=2 * w + 1, mode="nearest")

    # Release, without a per-sample Python loop over 19 million samples.
    #
    # A one-pole lowpass alone is unusable here: where `need` dips it lags, so
    # the applied gain stays ABOVE what the peak required and the ceiling is
    # breached exactly where it matters. Taking the elementwise minimum of the
    # raw requirement and its smoothed version fixes that by construction — the
    # result is never greater than `need`, so attack stays immediate, while after
    # a dip the smoothed branch is what recovers, slowly. That is a release.
    a = math.exp(-1.0 / (release * SR))
    smooth = lfilter([1.0 - a], [1.0, -a], need, zi=np.array([a * need[0]]))[0]
    return stereo * np.minimum(need, smooth)


def normalise(raw: np.ndarray, quiet: bool = False) -> np.ndarray:
    """To -16 LUFS with the true peak under -1 dBTP — both, not one of them.

    Two knobs, each moved only by measurement, and every pass starts from the
    unprocessed render so limiting never compounds on itself:

      gain   what it takes to hit the loudness target. Limiting costs a little
             loudness, so this is re-derived from the limited result.
      guard  how far BELOW the ceiling the limiter is actually set. The limiter
             works on sample peaks and the spec is on inter-sample peaks, which
             are strictly higher — the first master limited to -1.0 and delivered
             -0.69 dBTP. Rather than guess the overshoot, measure it and open the
             guard by exactly what was over.
    """
    measured = lufs(raw)
    if not math.isfinite(measured):
        return raw
    gain_db = TARGET_LUFS - measured
    guard = 0.3
    out = raw
    for _ in range(8):
        out = limit(raw * db_to_gain(gain_db), TRUE_PEAK_DBTP - guard)
        tp, loud = true_peak_db(out), lufs(out)
        if tp > TRUE_PEAK_DBTP:
            guard += (tp - TRUE_PEAK_DBTP) + 0.05
            continue
        if abs(loud - TARGET_LUFS) < 0.15:
            break
        gain_db += TARGET_LUFS - loud
    loud, tp = lufs(out), true_peak_db(out)
    if not quiet:
        flag = "" if abs(loud - TARGET_LUFS) < 0.5 and tp <= TRUE_PEAK_DBTP + 0.05 else "   <-- OFF SPEC"
        print(f"  {loud:.2f} LUFS · true peak {tp:.2f} dBTP{flag}")
    return out


def control_track(window: str, seed: int | None, rate: int) -> dict:
    cmd = ["node", str(CONTROL), "--window", window, "--rate", str(rate)]
    if seed is not None:
        cmd += ["--seed", str(seed)]
    done = subprocess.run(cmd, capture_output=True, text=True)
    if done.returncode != 0:
        sys.exit(f"control track failed:\n{done.stderr.strip()}")
    return json.loads(done.stdout)


def main() -> int:
    ap = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    ap.add_argument("--window", default="master", help="any window declared in render/program.json")
    ap.add_argument("--seed", help="override the program seed; accepts 0x notation")
    ap.add_argument("--rate", type=int, default=30, help="control-track sampling rate in Hz")
    ap.add_argument("--bank", type=Path, default=BANK)
    ap.add_argument("--out", type=Path, default=None)
    args = ap.parse_args()

    seed = int(args.seed, 0) if args.seed else None
    bank = Bank(args.bank)
    control = control_track(args.window, seed, args.rate)

    print(f"{control['title']} · {control['window']} · seed 0x{control['seed']:X} · {control['duration']:.1f}s")
    print(f"  bank {bank.fingerprint} · {len(bank.grains)} grains from {len(bank.sources)} recordings")

    stereo = normalise(render(control, bank))

    out = args.out or (OUT / f"{control['window']}-0x{control['seed']:X}.wav")
    out.parent.mkdir(parents=True, exist_ok=True)
    wavfile.write(out, SR, stereo.T.astype(np.float32))
    print(f"  {out}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
